### DISCUSSION TITLE
Will C# benefit from Computation Expressions?

### DISCUSSION THREAD

**Discussion by voroninp** [dotnet/csharplang#6777] [Q&A]
F# has [a very nice feature](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/computation-expressions) which makes code look nicer compared to what it would be if methods were used directly.

C# linq syntax with `from` and `let` is similar to CE, but it's far less expressive.

**Comment by svick** [comment]
The main uses of computation expressions in F# are addressed by LINQ, `async`-`await` and `yield return` in C#. In other words, where F# has a single flexible system for certain kinds of mini-DSLs, C# has dedicated syntax for each of the primary use cases.

This means if there is some niche use case for F# computation expressions, C# very likely isn't going to have a great syntax for it.

If there is some common use case that's missing, the best option is probably to expand one of the existing syntaxes.

Do you have a specific use in mind?

**Reply by voroninp** [reply]
Yes, for `Result<TSuccess, TError>`.

**Reply by svick** [reply]
C# already has a dedicated syntax for that: `try`-`catch`. I get that there are advantages to `Result` over exceptions, but it's not a common approach in C# and I don't think C# is going to introduce features specifically to make a different error-handling strategy more convenient.

Though C# is probably going to add discriminated unions in the future (see https://github.com/dotnet/csharplang/issues/113 and https://github.com/dotnet/csharplang/issues/3179), which could make working with `Result` nicer.

**Reply by Joe4evr** [reply]
Side-note: It would be useful if, with the support of DUs, there comes a BCL "blessed" version of `Result<T, TError>` just so that each library doesn't have to define their own and there's good interoperability.

**Reply by voroninp** [reply]
@svick true, it's not common, but this can also be a survivor bias. It's not convenient to use alternative approaches, so it's not common.

**Reply by HaloFour** [reply]
I've brought up this conversation on the Discord server before, specifically referencing Haskell do notation and Scala for comprehensions, and there didn't seem to be a lot of interest in such functionality for composing monads in this manner.  I'd be curious to see if DUs have an impact on that opinion, especially if the BCL gets a "blessed" `Result<T, TError>` monad.  I could see some language features to bridge the current world of exceptions, but I kinda doubt they'd do something this general purpose.

**Reply by Richiban** [reply]
> some common use case that's missing

I'd argue that the common use case is the ability to write your own. I think C# could invent new functionality that would combine F#'s computation expressions with Kotlin's builders; it would be awesome but probably quite niche.
